### DISCUSSION TITLE
Use of RSA_test_flags generates Segmentation fault

### DISCUSSION THREAD

**Discussion by debananda-pal** [openssl/openssl#24837] [Q&A]
Hi Sir/Madam,

We are using OpenSSL 3.0.11 as part of our embedded development on Ubuntu based OS.

While creating certificate we use X509_REQ_new().
But later after few more operations, we find segmentation fault as below :

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/bin/python3 -u /opt/dell/os10/lib/mgmtsys.py'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f13fca77440 in RSA_test_flags ()
   from /lib/x86_64-linux-gnu/libcrypto.so.3
[Current thread is 1 (Thread 0x7f13d37fe6c0 (LWP 2165))]
(gdb)
(gdb) bt
#0  0x00007f13fca77440 in RSA_test_flags ()
   from /lib/x86_64-linux-gnu/libcrypto.so.3
#1  0x00007f13fca721fe in ?? () from /lib/x86_64-linux-gnu/libcrypto.so.3
#2  0x00007f13fca722b0 in ?? () from /lib/x86_64-linux-gnu/libcrypto.so.3
#3  0x00007f13fcae9157 in X509_PUBKEY_set ()
   from /lib/x86_64-linux-gnu/libcrypto.so.3

Kindly give your suggestion to proceed on this.

Regards,
D Pal

**Comment by nhorman** [comment]
can you share the code you've written?

**Comment by debananda-pal** [comment]
Thank you nhorman.
Please find the pseudo code below and let me know your suggestion.
This code is working on Ubuntu Buster version, We are now using Bookworm version.

int add_req_ext(STACK_OF(X509_EXTENSION) *sk, int nid, const char *value)
{
    X509_EXTENSION *ex;
    char val[BUF_LEN];

    strncpy(val, value, sizeof(val));
    val[BUF_LEN-1] = '\0';

    ex = X509V3_EXT_conf_nid(NULL, NULL, nid, val);
    if (!ex) {
        EV_LOGGING(MGMT,ERR,"CMA","X509V3_EXT_conf_nid failed");
        return CMA_ERR_GEN;
    }
    sk_X509_EXTENSION_push(sk, ex);

    return CMA_NO_ERR;
}

int CSR-Cert-gen(const char *crtfile, const char *keyfile,
                const unsigned char *cname, const char *altname,
                const unsigned char *country, const unsigned char * state,
                const unsigned char *locality, const unsigned char *org,
                const unsigned char *orgunit, unsigned char *email,
                int keylength, char * status_detail, size_t status_size)
{
    int status = CMA_NO_ERR;
    FILE *crtfp = NULL;
    FILE *keyfp = NULL;
    X509_REQ *csr = NULL;
    EVP_PKEY *privateKey = NULL;
    X509_NAME *name = NULL;
    RSA *keyPair = NULL;
    BIGNUM *bigNumber = NULL;
    STACK_OF(X509_EXTENSION) *exts = NULL;
    struct stat sb;
    char *openSSLerrString = NULL, *reasonString = NULL;
    struct dirent *direntp;
    char pvtkeyfilename[BUF_LEN];
    DIR *dir_ptr = NULL;
    int max = 0, num = 0, rc = 0;
    char *ptr;
    EVP_PKEY_CTX *ctx;
    EVP_PKEY *pkey = NULL;
	
	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
	/*Error validation for ctx*/
	
	/*Validation of inputs crtfile, keyfile */
	
	crtfp = fopen(crtfile, "w");
	/*Error validation for crtfp*/
	
	keyfp = fopen(keyfile, "w");
	/*Error validation for keyfp*/
	
	csr = X509_REQ_new();
	
	
	name = X509_REQ_get_subject_name (csr);
	
	if (cname && *cname)
    {
        if (!X509_NAME_add_entry_by_txt (name, "CN", MBSTRING_ASC, cname, -1, -1, 0)) {
	    /*Error validation */
	}

	/* Input validation for orgunit, org, country, state, locality, email, */
	
	/* Add various extensions: standard extensions */
    exts = sk_X509_EXTENSION_new_null();
    status = add_req_ext(exts, NID_key_usage, "critical,digitalSignature,keyEncipherment");
	
	X509_REQ_add_extensions(csr, exts);
	sk_X509_EXTENSION_pop_free(exts, X509_EXTENSION_free);
	
	EVP_PKEY *keyPair = EVP_PKEY_new();
	// Validation
	
	bigNumber = BN_new();
	// Validation
	
	BN_set_word (bigNumber, RSA_F4);///
	
	// Generate the key pair; lots of computes here
    if (!EVP_PKEY_generate(ctx,&pkey));
	
	privateKey = EVP_PKEY_new();
	
	EVP_PKEY_set_type(privateKey, EVP_PKEY_RSA);
	
	keyPair = NULL;
	
	// Set the certificate's public key from the private key object
	X509_REQ_set_pubkey (csr, privateKey);   ------->>>>>>  Looks like getting Segmentation fault here
	
      // Sign it with SHA-256
     X509_REQ_sign (csr, privateKey, EVP_sha256());
	
     EVP_PKEY *evp_key = EVP_PKEY_new();
	  
      BIO *bio = BIO_new(BIO_s_mem());
      if (!bio || !PEM_write_bio_PrivateKey(bio, evp_key, NULL, NULL, 0, NULL, NULL)){
	      /*validation*/
      }
	
     PEM_write_X509_REQ(crtfp ,csr);
	
    fclose(crtfp);
    fclose(keyfp);

    BN_clear_free (bigNumber);
    EVP_PKEY_free (privateKey);
    X509_REQ_free (csr);
}

**Comment by debananda-pal** [comment]
Hi, please let me know if there is any suggestion.
Moreover I am seeing below APIs are deprecated in OpenSSL 3.0.11 :
RSA_new(), RSA_generate_key_ex(), EVP_PKEY_assign_RSA(), EVP_PKEY_get1_RSA(), PEM_write_RSAPrivateKey()
Above APIs are used to create RSA privatekey to assign to Self signed or CSR certificate.

Please let me know the corresponding APIs which are supported in OpenSSL 3.0.11 .

Otherwise, please let me know the APIs to use to create privatekey with OpenSSL 3.0.11

**Comment by nhorman** [comment]
Its a bit hard to tell due to the formatting above, but at a minimum you seem to be setting new keys to the certificate, rather than the ones you generated.

There are examples of how to use the apis you are after here:
https://www.openssl.org/docs/man3.3/man3/EVP_PKEY_generate.html
